test/cpp/qps/client_async.cc

```diff
@@ -151,191 +118,84 @@ class ClientRpcContextUnaryImpl : public ClientRpcContext {
   ResponseType response_;
   bool (ClientRpcContextUnaryImpl::*next_state_)();
   std::function<void(grpc::Status, ResponseType *)> callback_;
+  std::function<std::unique_ptr<grpc::ClientAsyncResponseReader<ResponseType>>(
+      TestService::Stub *, grpc::ClientContext *, const RequestType &, void *)>
+      start_req_;
   grpc::Status status_;
   double start_;
   std::unique_ptr<grpc::ClientAsyncResponseReader<ResponseType>>
       response_reader_;
 };
 
-static void RunTest(const int client_threads, const int client_channels,
-                    const int num_rpcs, const int payload_size) {
-  gpr_log(GPR_INFO,
-          "QPS test with parameters\n"
-          "enable_ssl = %d\n"
-          "client_channels = %d\n"
-          "client_threads = %d\n"
-          "num_rpcs = %d\n"
-          "payload_size = %d\n"
-          "server_host:server_port = %s:%d\n\n",
-          FLAGS_enable_ssl, client_channels, client_threads, num_rpcs,
-          payload_size, FLAGS_server_host.c_str(), FLAGS_server_port);
-
-  std::ostringstream oss;
-  oss << FLAGS_server_host << ":" << FLAGS_server_port;
-
-  class ClientChannelInfo {
-   public:
-    explicit ClientChannelInfo(const grpc::string &server)
-        : channel_(CreateTestChannel(server, FLAGS_enable_ssl)),
-          stub_(TestService::NewStub(channel_)) {}
-    ChannelInterface *get_channel() { return channel_.get(); }
-    TestService::Stub *get_stub() { return stub_.get(); }
+class AsyncClient GRPC_FINAL : public Client {
+ public:
+  explicit AsyncClient(const ClientConfig &config) : Client(config) {
+    for (int i = 0; i < config.async_client_threads(); i++) {
+      cli_cqs_.emplace_back(new CompletionQueue);
+    }
 
-   private:
-    std::shared_ptr<ChannelInterface> channel_;
-    std::unique_ptr<TestService::Stub> stub_;
-  };
+    auto payload_size = config.payload_size();
+    auto check_done = [payload_size](grpc::Status s, SimpleResponse *response) {
+      GPR_ASSERT(s.IsOk() && (response->payload().type() ==
+                              grpc::testing::PayloadType::COMPRESSABLE) &&
+                 (response->payload().body().length() ==
+                  static_cast<size_t>(payload_size)));
+    };
+
+    int t = 0;
+    for (int i = 0; i < config.outstanding_rpcs_per_channel(); i++) {
+      for (auto &channel : channels_) {
+        auto *cq = cli_cqs_[t].get();
+        t = (t + 1) % cli_cqs_.size();
+        auto start_req = [cq](TestService::Stub *stub, grpc::ClientContext *ctx,
+                              const SimpleRequest &request, void *tag) {
+          return stub->AsyncUnaryCall(ctx, request, cq, tag);
+        };
+
```

Comment:
I'm kind of ok with it being in the initiating loop - my main worry
initially was that it was part of the main loop.

On Wed, Mar 4, 2015 at 2:17 PM Vijay Pai <notifications@github.com> wrote:

> In test/cpp/qps/client_async.cc
> <https://github.com/grpc/grpc/pull/837#discussion_r25820106>:
>
> > +      GPR_ASSERT(s.IsOk() && (response->payload().type() ==
> > +                              grpc::testing::PayloadType::COMPRESSABLE) &&
> > +                 (response->payload().body().length() ==
> > +                  static_cast<size_t>(payload_size)));
> > +    };
> > +
> > +    int t = 0;
> > +    for (int i = 0; i < config.outstanding_rpcs_per_channel(); i++) {
> > +      for (auto &channel : channels_) {
> > +        auto *cq = cli_cqs_[t].get();
> > +        t = (t + 1) % cli_cqs_.size();
> > +        auto start_req = [cq](TestService::Stub *stub, grpc::ClientContext *ctx,
> > +                              const SimpleRequest &request, void *tag) {
> > +          return stub->AsyncUnaryCall(ctx, request, cq, tag);
> > +        };
> > +
>
> A question you had given me earlier; should this be moved outside the
> actual request initiating loop? If the only capture is for cq, maybe that
> could be a new argument for ClientRcpContextUnaryImpl and then get passed
> back as an argument to this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc/pull/837/files#r25820106>.
>